### PR TITLE
Updates to cudaextender ungapped_xdrop kernel

### PR DESCRIPTION
**Pull request by gsneha26**

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: cudaextender/src/ungapped_xdrop_kernels.cuh
Comment (ahehn-nv) [inline cudaextender/src/ungapped_xdrop_kernels.cuh:65]: I would use `const ScoredSegmentPair&` here.

**Comment by tijyojwad** [comment]
add to allowlist

File: cudaextender/src/ungapped_xdrop.cu
Comment (tijyojwad) [inline cudaextender/src/ungapped_xdrop.cu:182]: since this is unused code, feel free to remove it. if we need it we can always look in the git history

**Review by tijyojwad — CHANGES_REQUESTED** [review]

File: cudaextender/src/ungapped_xdrop_kernels.cuh
Comment (ahehn-nv) [inline cudaextender/src/ungapped_xdrop_kernels.cuh:77]: This looks a bit dangerous to me. If I understand it correctly, you add the full length of the extension to the target position, instead of just the right or left extension length.
This might be an artificial case, but assume I have
```
x.seed_pair.target_position_in_read = 30;
x.length = 5; (because it extends from 30-34)
y.seed_pair.target_position_in_read = 20;
y.length = 20; (because it extends from 1-20)
```
=> `(x.seed_pair.target_position_in_read >= y.seed_pair.target_position_in_read) = true` and `(x.seed_pair.target_position_in_read + x.length) <= (y.seed_pair.target_position_in_read + y.length) = true`
even though these are disjoint extensions.
Does this not matter or can it never happen?

File: cudaextender/src/ungapped_xdrop_kernels.cuh
Comment (ahehn-nv) [inline cudaextender/src/ungapped_xdrop_kernels.cuh:104]: If the shift/difference `(*.seed_pair.target_position_in_read - *.seed_pair.query_position_in_read)` is equal for `x` and `y`, and the `*.seed_pair.target_position_in_read` is equal for them, then `*.seed_pair.query_position_in_read` must be equal as well. Hence this comparison can be skipped.
Maybe it was meant to compare `x.length` and `y.length`? Or is that not needed anymore?

**Review by ahehn-nv — CHANGES_REQUESTED** [review]
Great! Thanks for the PR, Sneha! I've got a few questions and change requests. The ones starting with "Optional:" are just suggestions which you may choose to ignore. :)

File: cudaextender/src/ungapped_xdrop_kernels.cuh
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop_kernels.cuh:65]: Good point! I will change it.

**Comment by ahehn-nv** [comment]
I noticed, the code style check failed.
@gsneha26 : when you build with cmake, execute `make format` in the build directory to fix the code formatting automatically. It should pass the test then. (You might need to install "clang-format" for `make format` to work - most Linux distributions provide a package for it.)

File: cudaextender/data/sample_scored_segment_pairs.csv
Comment (atadkase) [inline cudaextender/data/sample_scored_segment_pairs.csv:1036]: Sorry I missed the test data change portion of the discussion. Was just the order changed? In that case, can you sort one of these (read scored_segment_pairs or output scored_segment_pairs) before equality checking?

**Comment by tijyojwad** [comment]
Hi @gsneha26 - friendly ping for the comments on this PR. Let us know if you have any clarifying questions or need help!

**Review by atadkase — CHANGES_REQUESTED** [review]

File: cudaextender/src/ungapped_xdrop_kernels.cuh
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop_kernels.cuh:77]: I don't think I understand correctly. But just for context -- Going from the seed pair to segment pair, `segment_pair.target_position_in_read = seed_pair.target_position_in_read - left_extension_length`. Also, for this equality comparison, we look at the segments pairs on the same diagonal i.e. `(x.seed_pair.target_position_in_read - x.seed_pair.query_position_in_read) == (y.seed_pair.target_position_in_read - y.seed_pair.query_position_in_read)`.

File: cudaextender/src/ungapped_xdrop.cu
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop.cu:39]: I had added those while making changes. I will remove those, thanks!

File: cudaextender/src/ungapped_xdrop.cu
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop.cu:182]: Will do, thanks!

File: cudaextender/src/ungapped_xdrop_kernels.cu
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop_kernels.cu:193]: Will do, thanks!

File: cudaextender/src/ungapped_xdrop_kernels.cu
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop_kernels.cu:365]: Will do, thanks!

File: cudaextender/data/sample_scored_segment_pairs.csv
Comment (gsneha26) [reply cudaextender/data/sample_scored_segment_pairs.csv:1036]: There have been changes to the data too. During the development after generating this first sample, I resolved the case where there are multiple segment pairs that differ by 1 in the positions but should they should map to the same final segment pair. For example,

```
24377,24377,202,18237
24377,24377,201,18146
24377,24377,200,18046
24377,24377,199,17955
24377,24377,198,17864
```

File: cudaextender/src/ungapped_xdrop_kernels.cu
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop_kernels.cu:255]: Good point! I have added `constexpr int32_t nuc_entropy = 4;` for replacing all `4` in entropy calculations.

**Comment by gsneha26** [comment]
@ahehn-nv thanks for the `make format` suggestion, that helped a lot. I have tried to address all the comments, do let me know if there is any other issue.

**Review by ahehn-nv — CHANGES_REQUESTED** [review]
Almost good to merge. :)

File: cudaextender/src/ungapped_xdrop_kernels.cuh
Comment (ahehn-nv) [reply cudaextender/src/ungapped_xdrop_kernels.cuh]: Hm. I'm still a bit confused about this one.
Could you explain me in words: when would you consider two ScoredSegmentPairs to be equal?
We can also hop on a short call if that makes it easier to explain.

File: cudaextender/src/ungapped_xdrop_kernels.cu
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop_kernels.cu:290]: Done!

File: cudaextender/src/ungapped_xdrop_kernels.cu
Comment (atadkase) [inline cudaextender/src/ungapped_xdrop_kernels.cu:360]: @ahehn-nv Since we're using bitwise shift operator here, would it be better to keep `offset` unsigned, though it might never reach the sign bit?

File: cudaextender/src/ungapped_xdrop_kernels.cu
Comment (atadkase) [inline cudaextender/src/ungapped_xdrop_kernels.cu:379]: [nit]
```suggestion
                const int32_t temp_score     = __shfl_up_sync(0xFFFFFFFF, max_thread_score, offset);
```

File: cudaextender/src/ungapped_xdrop_kernels.cu
Comment (atadkase) [inline cudaextender/src/ungapped_xdrop_kernels.cu:190]: Why is this qualifier necessary? `lane_id == 0` will never get these updates. Is that intended?

File: cudaextender/src/ungapped_xdrop_kernels.cu
Comment (ahehn-nv) [reply cudaextender/src/ungapped_xdrop_kernels.cu:360]: I understand your concern, but I'd keep `offset` signed, since it still represents a number, not a bit pattern or something. Shifting too far would be similar to an integer overflow by regular multiplication or addition.

File: cudaextender/src/ungapped_xdrop_kernels.cuh
Comment (ahehn-nv) [reply cudaextender/src/ungapped_xdrop_kernels.cuh]: We discussed offline and I understand what my misconception was now. Makes sense now. Thanks!

File: cudaextender/src/ungapped_xdrop_kernels.cuh
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop_kernels.cuh:70]: Difference between the target and query coordinates determines the "diagonal" on which the segment pair lies. 2 segment pairs can overlap only if they are on the same diagonal.

File: cudaextender/src/ungapped_xdrop_kernels.cu
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop_kernels.cu:190]: It is a prefix scan with OR operation (just like the earlier loops with AND and MAX operations). Hence, `lane_id == 0` should not be updated

**Review by ahehn-nv — APPROVED** [review]
Looks good to me now.

File: cudaextender/README.md
Comment (ahehn-nv) [reply cudaextender/README.md]: Discussed offline - will use https://doi.ieeecomputersociety.org/10.1109/SC41405.2020.00043 .

File: cudaextender/src/ungapped_xdrop_kernels.cuh
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop_kernels.cuh:94]: Done

File: cudaextender/src/ungapped_xdrop.cu
Comment (gsneha26) [reply cudaextender/src/ungapped_xdrop.cu:165]: Sorry I don't understand, could you elaborate?

**Comment by gsneha26** [comment]
@ahehn-nv - the checks seem to be failing due to some 500 error, could you maybe suggest what to do? Thanks!

**Comment by ahehn-nv** [comment]
rerun tests

**Comment by ahehn-nv** [comment]
There seems to be a problem with the test system. I'll take care of it and get back to you.

**Comment by ahehn-nv** [comment]
rerun tests

**Comment by gsneha26** [comment]
What is the best way to trigger the checks again?

**Comment by ahehn-nv** [comment]
Write the same comment here as I did above, or push a new commit.

**Comment by ahehn-nv** [comment]
But all checks have passed now.

**Comment by gsneha26** [comment]
Great! Sorry I didn't check. Thanks @ahehn-nv

**Comment by ahehn-nv** [comment]
@gsneha26 Welcome. :) Oh now I understand. Sorry, that comment wasn't meant for you, but for the test system to rerun its stuff. :) As I said, all checks pass now, so we just need @atadkase's approval now. :+1:

File: cudaextender/src/ungapped_xdrop.cu
Comment (atadkase) [reply cudaextender/src/ungapped_xdrop.cu:165]: Requested change for lines 77-93 in this file:
```
/Figure out memory requirements for cub functions
    size_t cub_storage_bytes  = 0;
    GW_CU_CHECK_ERR(cub::DeviceSelect::Unique(nullptr,
                                              cub_storage_bytes,
                                              d_tmp_ssp_.data(),
                                              d_tmp_ssp_.data(),
                                              (int32_t*)nullptr,
                                              batch_max_ungapped_extensions_,
                                              stream_));
```

File: cudaextender/src/ungapped_xdrop.cu
Comment (atadkase) [inline cudaextender/src/ungapped_xdrop.cu:34]:
```suggestion
```

**Review by tijyojwad — APPROVED** [review]

File: cudaextender/src/ungapped_xdrop.cu
Comment (atadkase) [inline cudaextender/src/ungapped_xdrop.cu:77]: Thanks for making this change. Please feel free to merge this after removing the temp_storage_bytes unused var.

**Review by atadkase — APPROVED** [review]
Approved - one small change requested (left a comment at that spot).

**Comment by ahehn-nv** [comment]
rerun tests

**Comment by gsneha26** [comment]
I think everything has been worked out. Thanks @tijyojwad @ahehn-nv @atadkase . I do not have write access to merge the commits.

**Comment by ahehn-nv** [comment]
Thanks for your PR, @gsneha26! I'll merge it.

**Pull request merged**
